### ISSUE TITLE
10x drop in performance from 4.1.0 to 4.2.0

### ISSUE THREAD

**Issue by alamothe** [js-joda/js-joda#573]
Hello,
We recently upgraded from:
```
    "js-joda": "1.10.1",
    "js-joda-timezone": "2.0.2",
```
to:
```
    "@js-joda/core": "^4.3.1",
    "@js-joda/timezone": "^2.10.0",
```
We experience over 10x drop in performance. We run on Node 14. 

Do you know why this could be the case?

**Comment by alamothe** [comment]
I pinpointed the crazy 10x drop to `4.1.0` -> `4.2.0`.

The former is fine, the latter is busted.

**Comment by pithu** [comment]
Hi @alamothe, thanks for tracking this down. What is your test case for performance measurement, so we can check that ?

**Comment by alamothe** [comment]
Hello, our use case is a production use case, basically using a lot of datetime calculation with timezones during a single request. In `4.1.0` (and all versions before it) it takes about 1s total, but in `4.2.0` (and all versions after it) it takes 12s and CPU and memory usage is unstable during that time.

Does your benchmarks capture anything like that? If not, could the problem be in the bundle itself?

**Comment by pithu** [comment]
Actually the changes between `@js-joda/core` `4.1.0` -> `4.2.0` are formatting and code style, no functional changes, that's a bit strange. What versions of `@js-joda/timezone` are you comparing ? 
And some code snippets would help to track the issue down, so that we can do some benchmark tests.

**Comment by alamothe** [comment]
I checked the distribution and it uses a completely different approach to generate classes. Could that be the problem? 

Do you have end-to-end tests that run on the actual package? If not, I can try to write something.

**Comment by alamothe** [comment]
I have a strong suspicion that this is the culprit: github.com/js-joda/js-joda/commit/38ada022d35d70055e4b5e872482438149a6dbe1

It completely changed the generated file.

**Comment by pithu** [comment]
@alamothe thanks again for reporting this. I can reproduce the issue

We have a benchmark test in the package, i updated it and moved it into the example package. See #574 , i also added the result logs for 4.2.1 and 4.1.0 to the pr, that are showing the dramatic performance drop.

I guess the reason is, that we changed the babel targets (mainly for ie11) https://github.com/js-joda/js-joda/commit/38ada022d35d70055e4b5e872482438149a6dbe1#diff-5f0b6a82459cf8b0d57cd099f277da5bca901d8e8469486fa5a53ee7648d9365

I will investigate further

**Comment by InExtremaRes** [comment]
Hi there @pithu.

Have you tested the performance of not transpiling down to ES5?

This is something I've been thinking about. A particular engine _might_ be able to better optimize more modern syntax constructs (if not right now, maybe in the future) since those tend to be more intention revealing. Thus providing two different bundles could be beneficial; I bet most people can run at least ES2017, so no need to downlevel classes and loops for them. A different bundle can be provided for those stuck in ES5.

Is this something you have considered?

**Comment by alamothe** [comment]
Not an expert at all, but if people use bundlers I believe they will transpile the library according to their browserlist, so you don't have to do anything.

**Comment by pithu** [comment]
@InExtremaRes as already discussed, i would like to prevent maintaining several build outputs, if possible.

**Comment by pithu** [comment]
@alamothe, we are providing the code as transpiled UMD and ESM module. Usually you are not transpiling third party code in node_modules and in browser environments global vars are still common, therefore i would say with UMD modules we are on the right track. Its running out of the box in backend and frontend.

Reverting the babel targets, seems to fix your issue see #574 . Could you confirm this ?

**Comment by pithu** [comment]
should be fixed with `@js-joda/core:5.0.0`

**Comment by pithu** [comment]
I close this issue for now, please feel free to reopen it if there are still questions or issues
